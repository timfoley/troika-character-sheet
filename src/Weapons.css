.weapons {
	grid-area: weapons;
	display: flex;
	flex-direction: column;
}

.weapons h2 {
	margin-top: 0;
}

.weapons__grid {
	display: grid;
	grid-gap: 2px;
	row-gap: 12px;
	grid-template-columns: 7fr repeat(7, 1fr);
}

@media (min-width: 500px) {
	.weapons__grid {
		display: grid;
		grid-gap: 8px;
		row-gap: 12px;
		grid-template-columns: 1fr repeat(7, 26px);
	}

	.weapon__damage {
		height: 24px;
	}
}

.weapon__name > input {
	width: 100%;
	border: none;
	border-bottom: 1px solid black;
	margin-bottom: 0;
	border-radius: 0;
}

.weapon__damage {
	margin-top: auto;
	border: 1px solid black;
	position: relative;
}

.weapon__damage input {
	width: 100%;
	height: 100%;
	padding: 0;
	margin: 0;
	border: none;
	display: flex;
	justify-content: center;
	align-items: center;
	text-align: center;
	font-weight: bold;
}

.weapon__damage__roll {
	position: absolute;
	top: -15px;
	left: 8px;
	font-size: 0.7rem;
}

.weapons__add-button {
	margin-left: auto;
	font-size: 0.8rem;
	margin-top: 8px;
}
