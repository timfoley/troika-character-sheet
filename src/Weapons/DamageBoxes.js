import React from "react";

export function DamageBoxes(props) {
	function getRollFromIndex(i) {
		if (i < 6) {
			return (i + 1).toString();
		}

		return "7+";
	}

	return props.damage.map((dmgValue, i) => (
		<div className="weapon__damage" key={props.name + "-dmg-" + i}>
			<input
				className={`weapon__damage__input weapon__damage__input--${props.wpnIndex} hide-spinners`}
				value={dmgValue}
				aria-label={`${getRollFromIndex(i)} damage value for ${
					props.name
				}`}
				type="number"
				onChange={e =>
					props.setWeaponDamage(
						i,
						parseInt(e.target.value),
						props.wpnIndex
					)
				}
				onBlur={props.updateCharacter}
			/>
			<div className="weapon__damage__roll" aria-hidden="true">
				{getRollFromIndex(i)}
			</div>
		</div>
	));
}
