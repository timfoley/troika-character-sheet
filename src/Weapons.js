import React from "react";
import PropTypes from "prop-types";

import "./Weapons.css";

function DamageBoxes(props) {
	function getRollFromIndex(i) {
		if (i < 6) {
			return (i + 1).toString();
		}

		return "7+";
	}

	return props.damage.map((dmgValue, i) => (
		<div className="weapon__damage" key={props.name + "-dmg-" + i}>
			<input
				className="hide-spinners"
				value={dmgValue}
				aria-label={`${getRollFromIndex(i)} damage value for ${
					props.name
				}`}
				type="number"
				onChange={e =>
					props.setWeaponDamage(
						i,
						parseInt(e.target.value),
						props.wpnIndex
					)
				}
				onBlur={props.updateCharacter}
			/>
			<div className="weapon__damage__roll" aria-hidden="true">
				{getRollFromIndex(i)}
			</div>
		</div>
	));
}

export function Weapons(props) {
	const addWeapon = () => {
		props.updateCharacter("weapons", [
			...props.weapons,
			{ name: "", damage: [0, 0, 0, 0, 0, 0, 0] },
		]);
	};

	const handleAddSkillKeyPress = e => {
		if (e.which === 32 || e.key === "enter") {
			addWeapon();
		}
	};

	const setWeaponName = (index, value) => {
		const newWeapons = props.weapons;
		newWeapons[index] = {
			...newWeapons[index],
			name: value,
		};
		props.updateCharacter("weapons", newWeapons);
	};

	const setWeaponDamage = (dmgIndex, value, wpnIndex) => {
		const newWeapons = props.weapons;
		const newDamage = newWeapons[wpnIndex].damage;
		newDamage[dmgIndex] = isNaN(value) ? null : value;

		newWeapons[wpnIndex].damage = newDamage;
		props.updateCharacter("weapons", newWeapons);
	};

	return (
		<div className="weapons">
			<h2>Weapons</h2>
			<div className="weapons__grid">
				{props.weapons.map((weapon, index) => (
					<React.Fragment key={"weapon" + index}>
						<div className="weapon__name">
							<input
								aria-label="weapon name"
								type="text"
								defaultValue={weapon.name}
								onChange={e =>
									setWeaponName(index, e.target.value)
								}
							/>
						</div>
						<DamageBoxes
							damage={weapon.damage}
							name={weapon.name}
							setWeaponDamage={setWeaponDamage}
							wpnIndex={index}
						/>
					</React.Fragment>
				))}
			</div>
			<div
				onClick={addWeapon}
				onKeyPress={handleAddSkillKeyPress}
				tabIndex="0"
				className="weapons__add-button clickable"
			>
				Add Weapon
			</div>
		</div>
	);
}

Weapons.propTypes = {
	weapons: PropTypes.array,
	updateCharacter: PropTypes.func,
};

Weapons.defaultProps = {
	weapons: [{ name: "", damage: [0, 0, 0, 0, 0, 0, 0] }],
};
