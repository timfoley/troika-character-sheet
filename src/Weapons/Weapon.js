import React, { useEffect, useRef } from "react";
import PropTypes from "prop-types";

import { DamageBoxes } from "./DamageBoxes";

export const Weapon = props => {
	const weaponNameInput = useRef(null);

	useEffect(() => {
		props.shouldFocus && weaponNameInput.current.focus();
	}, [props.shouldFocus]);

	const setWeaponName = e => {
		const newWeapons = props.weapons;
		newWeapons[props.index] = {
			...newWeapons[props.index],
			name: e.target.value,
		};
		props.updateCharacter("weapons", newWeapons);
	};

	const setWeaponDamage = (dmgIndex, value, wpnIndex) => {
		const newWeapons = props.weapons;
		const newDamage = newWeapons[wpnIndex].damage;
		newDamage[dmgIndex] = isNaN(value) ? null : value;

		newWeapons[wpnIndex].damage = newDamage;
		props.updateCharacter("weapons", newWeapons);
	};

	return (
		<React.Fragment>
			<div className="weapon__name">
				<input
					className="weapon__name__input"
					aria-label="weapon name"
					type="text"
					value={props.weapon.name}
					onChange={setWeaponName}
					ref={weaponNameInput}
				/>
			</div>
			<DamageBoxes
				damage={props.weapon.damage}
				name={props.weapon.name}
				setWeaponDamage={setWeaponDamage}
				wpnIndex={props.index}
			/>
		</React.Fragment>
	);
};

Weapon.propTypes = {
	weapon: PropTypes.object.isRequired,
	updateCharacter: PropTypes.func.isRequired,
	weapons: PropTypes.array.isRequired,
	shouldFocus: PropTypes.bool,
};
