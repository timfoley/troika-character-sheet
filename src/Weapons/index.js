import React, { useState } from "react";
import PropTypes from "prop-types";

import "./Weapons.css";

import { Weapon } from "./Weapon";

export function Weapons(props) {
	const [indexToFocus, setIndexToFocus] = useState();

	const addWeapon = () => {
		setIndexToFocus(props.weapons.length);
		props.updateCharacter("weapons", [
			...props.weapons,
			{ name: "", damage: [0, 0, 0, 0, 0, 0, 0] },
		]);
	};

	const handleAddSkillKeyPress = e => {
		if (e.which === 32 || e.key === "enter") {
			addWeapon();
		}
	};

	return (
		<div className="weapons">
			<h2>Weapons</h2>
			<div className="weapons__grid">
				{props.weapons.map((weapon, index) => (
					<Weapon
						key={"weapon" + index}
						weapon={weapon}
						index={index}
						updateCharacter={props.updateCharacter}
						weapons={props.weapons}
						shouldFocus={indexToFocus === index}
					/>
				))}
			</div>
			<div
				onClick={addWeapon}
				onKeyPress={handleAddSkillKeyPress}
				tabIndex="0"
				className="weapons__add-button clickable"
			>
				Add Weapon
			</div>
		</div>
	);
}

Weapons.propTypes = {
	weapons: PropTypes.array,
	updateCharacter: PropTypes.func.isRequired,
};

Weapons.defaultProps = {
	weapons: [{ name: "", damage: [0, 0, 0, 0, 0, 0, 0] }],
};
