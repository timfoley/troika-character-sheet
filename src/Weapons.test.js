import React from "react";
import ReactDOM from "react-dom";
import { render } from "@testing-library/react";

import { Weapons } from "./Weapons";

it("renders Weapons without crashing", () => {
	const div = document.createElement("div");
	ReactDOM.render(<Weapons />, div);
	ReactDOM.unmountComponentAtNode(div);
});

it("matches snapshot", () => {
	const { container } = render(<Weapons />);
	expect(container.firstChild).toMatchSnapshot();
});

// it("renders a heading called 'Weapons'", () => {
//     const div = document.createElement("div");
//     ReactDOM.render(<Weapons />, div);
//     const { getByLabelText } = getQueriesForElement(div);
//     const input = getByLabelText(/name/i);
//     expect(input).toHaveAttribute("type", "text");
// })
