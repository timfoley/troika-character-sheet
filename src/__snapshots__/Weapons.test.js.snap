// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="weapons"
>
  <h2>
    Weapons
  </h2>
  <div
    class="weapons__grid"
  >
    <div
      class="weapon__name"
    >
      <input
        aria-label="weapon name"
        class="weapon__name__input"
        type="text"
        value=""
      />
    </div>
    <div
      class="weapon__damage"
    >
      <input
        aria-label="1 damage value for "
        class="weapon__damage__input weapon__damage__input--0 hide-spinners"
        type="number"
        value="0"
      />
      <div
        aria-hidden="true"
        class="weapon__damage__roll"
      >
        1
      </div>
    </div>
    <div
      class="weapon__damage"
    >
      <input
        aria-label="2 damage value for "
        class="weapon__damage__input weapon__damage__input--1 hide-spinners"
        type="number"
        value="0"
      />
      <div
        aria-hidden="true"
        class="weapon__damage__roll"
      >
        2
      </div>
    </div>
    <div
      class="weapon__damage"
    >
      <input
        aria-label="3 damage value for "
        class="weapon__damage__input weapon__damage__input--2 hide-spinners"
        type="number"
        value="0"
      />
      <div
        aria-hidden="true"
        class="weapon__damage__roll"
      >
        3
      </div>
    </div>
    <div
      class="weapon__damage"
    >
      <input
        aria-label="4 damage value for "
        class="weapon__damage__input weapon__damage__input--3 hide-spinners"
        type="number"
        value="0"
      />
      <div
        aria-hidden="true"
        class="weapon__damage__roll"
      >
        4
      </div>
    </div>
    <div
      class="weapon__damage"
    >
      <input
        aria-label="5 damage value for "
        class="weapon__damage__input weapon__damage__input--4 hide-spinners"
        type="number"
        value="0"
      />
      <div
        aria-hidden="true"
        class="weapon__damage__roll"
      >
        5
      </div>
    </div>
    <div
      class="weapon__damage"
    >
      <input
        aria-label="6 damage value for "
        class="weapon__damage__input weapon__damage__input--5 hide-spinners"
        type="number"
        value="0"
      />
      <div
        aria-hidden="true"
        class="weapon__damage__roll"
      >
        6
      </div>
    </div>
    <div
      class="weapon__damage"
    >
      <input
        aria-label="7+ damage value for "
        class="weapon__damage__input weapon__damage__input--6 hide-spinners"
        type="number"
        value="0"
      />
      <div
        aria-hidden="true"
        class="weapon__damage__roll"
      >
        7+
      </div>
    </div>
  </div>
  <div
    class="weapons__add-button clickable"
    tabindex="0"
  >
    Add Weapon
  </div>
</div>
`;
